packages/eslint-plugin/tests/rules/no-unnecessary-condition.test.ts

```diff
@@ -507,6 +507,53 @@ type Key = 'bar' | 'foo' | 'baz';
 declare const foo: Foo;
 declare const key: Key;
 
+foo?.[key]?.trim();
+    `,
+    // https://github.com/typescript-eslint/typescript-eslint/issues/7700
+    `
+type BrandedKey = string & { __brand: string };
+type Foo = { [key: BrandedKey]: string } | null;
+declare const foo: Foo;
+const key = '1' as BrandedKey;
+foo?.[key]?.trim();
+    `,
+    `
+type BrandedKey<S extends string> = S & { __brand: string };
+type Foo = { [key: string]: string; foo: 'foo'; bar: 'bar' } | null;
+type Key = BrandedKey<'bar'> | BrandedKey<'foo'>;
+declare const foo: Foo;
+declare const key: Key;
+foo?.[key].trim();
+    `,
+    `
+type BrandedKey = string & { __brand: string };
+interface Outer {
+  inner?: {
+    [key: BrandedKey]: string | undefined;
+  };
+}
+function Foo(outer: Outer, key: BrandedKey): number | undefined {
+  return outer.inner?.[key]?.charCodeAt(0);
+}
+    `,
+    `
+interface Outer {
+  inner?: {
+    [key: string & { __brand: string }]: string | undefined;
+    bar: 'bar';
+  };
+}
+type Foo = 'foo' & { __brand: string };
+function Foo(outer: Outer, key: Foo): number | undefined {
+  return outer.inner?.[key]?.charCodeAt(0);
+}
+    `,
+    `
+type BrandedKey<S extends string> = S & { __brand: string };
+type Foo = { [key: string]: string; foo: 'foo'; bar: 'bar' } | null;
+type Key = BrandedKey<'bar'> | BrandedKey<'foo'> | BrandedKey<'baz'>;
+declare const foo: Foo;
+declare const key: Key;
```

Comment:
[Tests] The linked issue runs with `noUncheckedIndexedAccess`, but these added valid cases don't exercise it. Could you add a couple of tests with it on please?

* [Playground showing false reports with `noUncheckedIndexedAccess` on](https://typescript-eslint.io/play/#ts=5.0.2&fileType=.tsx&code=C4TwDgpgBAQgTgQwHYBMIoNIRFAvFAZ2DgEskBzKAMigG8oB9AI0VQC4oBXVCAMzPRQAvgG4AUGLLAIcXggDG0AJJIkMgPJMAVhHnA6YqFDJq4HeMjSZs4oRPkB7JEShM8UAOQEHAW2gBrbA8oBAJYVissEHE0eQAbBDhoR2d9eQ4AJV0HOBQAHgseaxAAGigVU00dPQA%2BKAAfKCROOLjxMXkAfgA6AG0mAF0ekxkRIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRgA2uGw5E0aB2iRVEALoBfEIaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYAVSLQ%2Bg9ABMAkkVXoAHmoCC0aOhLCwo9FIC%2BIK0A&tokens=false)
* [Playground showing no false reports with `noUncheckedIndexedAccess` off](https://typescript-eslint.io/play/#ts=5.0.2&fileType=.tsx&code=C4TwDgpgBAQgTgQwHYBMIoNIRFAvFAZ2DgEskBzKAMigG8oB9AI0VQC4oBXVCAMzPRQAvgG4AUGLLAIcXggDG0AJJIkMgPJMAVhHnA6YqFDJq4HeMjSZs4oRPkB7JEShM8UAOQEHAW2gBrbA8oBAJYVissEHE0eQAbBDhoR2d9eQ4AJV0HOBQAHgseaxAAGigVU00dPQA%2BKAAfKCROOLjxMXkAfgA6AG0mAF0ekxkRIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRgA2uGw5E0aB2iRVEALoBfEIaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

I know it's probably not impactful here, but just in case.